Invalidate cache for environment variable and included file changes

devongovett opened the pull request:
Depends on #4963. Fixes #3708. Fixes #4264. Fixes #5132.

This invalidates the transformation cache when environment variables that are used in the code are updated, or when included files change. It adds an `env` node type to the request graph to track environment variables, and invalidates requests connected to changed vars on startup. Included files now store a hash as part of the `File` node, which is computed while the asset is being committed. Both of these are also added the the asset's `outputHash`, which is used to compute the hash of the asset graph and also in the packager cache key.

Actual invalidation of the transformer cache based on these is a bit tricky, and I'm not entirely sure I have the right solution here. Because we don't (and can't) know what the used environment variables and included files will be until after transforming the file, they can't be part of the cache key. Therefore, since they are already stored as part of the asset itself, we read the cache entry, and then check the referenced file hashes and environment variables against the current values. If they do not match, we don't use the cache entry and instead re-run the transformation pipeline. This works well for local caches, but could potentially cause multiple instances with different vars/included files overwriting each other in a shared cache.

An alternative would be to somehow send the environment variables and included files from the request graph to the transformer when available, and have it use these values as part of the cache key. But unfortunately, this would invalidate more than needed potentially, because intermediary assets that are created during transformation are not stored in the graph, and are only available in the cache itself. For example, for an HTML file with inline CSS, and JS that uses an environment variable, only the JS depends on the environment variable, not the HTML or CSS, so if the environment variable was part of the cache key for all of these assets, it would invalidate more than needed. Same for simpler cases, e.g. TS -> JS. The TS compilation doesn't need to be re-computed, just the JS pipeline. 🤔

parcel-benchmark commented:
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.63s | -9.00ms |
| Cached | 1.45s | -24.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/index.html | 701.00b | +0.00b | 1.72s | +119.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 1.71s | +125.00ms ⚠️ |
| dist/legacy/index.b3acc8eb.css | 77.00b | +0.00b | 1.57s | +146.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 30.00ms | -4.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 31.00ms | -3.00ms 🚀 |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 31.08s | -638.00ms |
| Cached | 1.74s | +119.00ms ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 47.00ms | -828.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 478.68kb | +0.00b | 75.00ms | -18.00ms 🚀 |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 11.00ms | -7.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.13m | +1.86s |
| Cached | 6.69s | +2.04s ⚠️ |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/popup.6c453e3c.js | 169.96kb | +0.00b | 54.52s | +20.00s ⚠️ |
| dist/EmojiPickerComponent.ae601247.js | 75.29kb | +0.00b | 9.94s | -1.51s 🚀 |
| dist/media-viewer.bcba38b3.js | 71.96kb | +0.00b | 14.76s | -1.98s 🚀 |
| dist/card.645a5c17.js | 55.01kb | +0.00b | 11.95s | -1.77s 🚀 |
| dist/card.0313bac8.js | 52.78kb | +0.00b | 56.52s | -42.04s 🚀 |
| dist/Modal.19ffed32.js | 41.27kb | +0.00b | 5.96s | -1.00s 🚀 |
| dist/component.d7c29980.js | 30.70kb | +0.00b | 2.90s | -162.00ms 🚀 |
| dist/component.54080a26.js | 22.50kb | +0.00b | 5.96s | -1.00s 🚀 |
| dist/DatePicker.0df05dac.js | 21.21kb | +0.00b | 9.93s | -1.49s 🚀 |
| dist/EmojiPickerComponent.7dc0ebd0.js | 18.23kb | +0.00b | 9.93s | -1.49s 🚀 |
| dist/js.5733a017.js | 16.40kb | +0.00b | 5.95s | -1.00s 🚀 |
| dist/ui.fee77548.js | 13.97kb | +0.00b | 11.91s | -1.69s 🚀 |
| dist/workerHasher.1e706e77.js | 11.75kb | +0.00b | 11.92s | -1.69s 🚀 |
| dist/component.77c0f682.js | 6.05kb | +0.00b | 6.01s | -1.26s 🚀 |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 14.71s | -2.02s 🚀 |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 14.72s | -2.02s 🚀 |
| dist/EmojiPickerComponent.627aca6b.js | 3.58kb | +0.00b | 9.93s | -1.49s 🚀 |
| dist/png-chunks-extract.a844fdb6.js | 3.41kb | +0.00b | 5.96s | -1.00s 🚀 |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 29.10s | -2.98s 🚀 |
| dist/Modal.0304b9ed.js | 3.08kb | +0.00b | 5.95s | -1.01s 🚀 |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 29.10s | -2.96s 🚀 |
| dist/ResourcedEmojiComponent.abe8596c.js | 2.00kb | +0.00b | 4.85s | -1.37s 🚀 |
| dist/date.231500d2.js | 1.73kb | +0.00b | 6.45s | -1.05s 🚀 |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 6.34s | -1.01s 🚀 |
| dist/images.80c92f5e.js | 1.67kb | +0.00b | 6.53s | -1.10s 🚀 |
| dist/16.1b6ca79f.js | 1.66kb | +0.00b | 7.41s | -1.08s 🚀 |
| dist/feedback.d5590b34.js | 1.64kb | +0.00b | 6.96s | -1.13s 🚀 |
| dist/workerHasher.fa198233.js | 1.54kb | +0.00b | 56.70s | -41.90s 🚀 |
| dist/list-number.c35f9589.js | 1.46kb | +0.00b | 6.62s | -1.12s 🚀 |
| dist/status.8ab4226b.js | 1.46kb | +0.00b | 6.80s | -1.17s 🚀 |
| dist/16.9ccb5b28.js | 1.38kb | +0.00b | 6.17s | -996.00ms 🚀 |
| dist/code.3dd71280.js | 1.38kb | +0.00b | 6.45s | -1.05s 🚀 |
| dist/16.19c93285.js | 1.33kb | +0.00b | 7.41s | -1.08s 🚀 |
| dist/16.c7751e34.js | 1.33kb | +0.00b | 6.17s | -997.00ms 🚀 |
| dist/16.18344f47.js | 1.31kb | +0.00b | 6.46s | -1.05s 🚀 |
| dist/link.2f4427e6.js | 1.30kb | +0.00b | 6.62s | -1.12s 🚀 |
| dist/heading6.fdbbdf4d.js | 1.30kb | +0.00b | 6.96s | -1.13s 🚀 |
| dist/heading3.2113b7a1.js | 1.29kb | +0.00b | 6.80s | -1.17s 🚀 |
| dist/16.2c4c1846.js | 1.27kb | +0.00b | 6.34s | -48.78s 🚀 |
| dist/emoji.93ae61d3.js | 1.23kb | +0.00b | 6.53s | -1.10s 🚀 |
| dist/16.d17ab186.js | 1.23kb | +0.00b | 6.34s | -1.01s 🚀 |
| dist/16.b4da0acb.js | 1.19kb | +0.00b | 6.34s | -1.01s 🚀 |
| dist/16.02c68c4f.js | 1.18kb | +0.00b | 6.46s | -1.05s 🚀 |
| dist/heading5.e195aab1.js | 1.17kb | +0.00b | 6.96s | -1.13s 🚀 |
| dist/expand.450f3aea.js | 1.16kb | +0.00b | 6.96s | -1.14s 🚀 |
| dist/media-card-analytics-error-boundary.2323af57.js | 1.15kb | +0.00b | 29.10s | -2.97s 🚀 |
| dist/16.4e3b0235.js | 1.14kb | +0.00b | 58.08s | +50.73s ⚠️ |
| dist/16.9ba230ea.js | 1.13kb | +0.00b | 6.17s | -997.00ms 🚀 |
| dist/16.a37df53b.js | 1.13kb | +0.00b | 6.17s | -996.00ms 🚀 |
| dist/16.9adb4ec7.js | 1.13kb | +0.00b | 6.17s | -996.00ms 🚀 |
| dist/16.eafe1a11.js | 1.13kb | +0.00b | 7.41s | -1.08s 🚀 |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 6.80s | -1.17s 🚀 |
| dist/16.3920766f.js | 1.10kb | +0.00b | 4.85s | -1.37s 🚀 |
| dist/component.6600f83e.js | 1.07kb | +0.00b | 4.85s | -1.37s 🚀 |
| dist/mention.4d64e059.js | 1.07kb | +0.00b | 6.62s | -1.12s 🚀 |
| dist/heading4.ec980e93.js | 1.06kb | +0.00b | 6.96s | -1.13s 🚀 |
| dist/16.afcb09b6.js | 1.05kb | +0.00b | 6.34s | -1.01s 🚀 |
| dist/layout.52ba15f7.js | 1.05kb | +0.00b | 6.53s | -1.10s 🚀 |
| dist/divider.16defde6.js | 1.03kb | +0.00b | 6.53s | -1.10s 🚀 |
| dist/quote.aa25b83c.js | 1.02kb | +0.00b | 6.71s | -1.14s 🚀 |
| dist/action.9aa72b39.js | 1.00kb | +0.00b | 6.45s | -1.05s 🚀 |
| dist/decision.20244046.js | 1010.00b | +0.00b | 6.53s | -1.10s 🚀 |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 6.71s | -1.14s 🚀 |
| dist/list.c9147a62.js | 978.00b | +0.00b | 6.62s | -1.12s 🚀 |
| dist/heading1.7d899ba5.js | 976.00b | +0.00b | 6.80s | -1.17s 🚀 |
| dist/panel-error.8d11e8b7.js | 906.00b | +0.00b | 6.62s | -1.12s 🚀 |
| dist/panel.a28c56ee.js | 897.00b | +0.00b | 6.71s | -1.14s 🚀 |
| dist/table.5a4874ee.js | 889.00b | +0.00b | 6.80s | -1.17s 🚀 |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 6.71s | -1.14s 🚀 |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 6.71s | -1.14s 🚀 |
| dist/media-card-analytics-error-boundary.663369b9.js | 835.00b | +0.00b | 11.91s | -1.74s 🚀 |
| dist/media-card-analytics-error-boundary.73ba4074.js | 835.00b | +0.00b | 56.70s | -41.90s 🚀 |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 14.76s | -2.00s 🚀 |
| dist/EmojiPickerComponent.2e54b9c6.js | 535.00b | +0.00b | 9.93s | -1.49s 🚀 |
| dist/simpleHasher.16c8bf41.js | 528.00b | +0.00b | 11.91s | -1.70s 🚀 |
| dist/simpleHasher.a29b2a8e.js | 528.00b | +0.00b | 56.89s | -41.71s 🚀 |
| dist/index.html | 119.00b | +0.00b | 3.56s | -371.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.474ded24.js | 2.72mb | -2.00b 🚀 | 146.00ms | -20.00ms 🚀 |
| dist/pdfRenderer.15a6a023.js | 1.11mb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/editorView.e06544fb.js | 587.12kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/popup.6c453e3c.js | 169.96kb | +0.00b | 140.00ms | -14.00ms 🚀 |
| dist/Toolbar.83ac0b6e.js | 92.34kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/EmojiPickerComponent.ae601247.js | 75.29kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/media-viewer.bcba38b3.js | 71.96kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/card.645a5c17.js | 55.01kb | +0.00b | 141.00ms | -14.00ms 🚀 |
| dist/card.0313bac8.js | 52.78kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/Modal.19ffed32.js | 41.27kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/component.d7c29980.js | 30.70kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/component.54080a26.js | 22.50kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/DatePicker.0df05dac.js | 21.21kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/EmojiPickerComponent.7dc0ebd0.js | 18.23kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/dropzone.7c704bdb.js | 17.27kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/smartMediaEditor.9623570c.js | 16.87kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/js.5733a017.js | 16.40kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/ui.fee77548.js | 13.97kb | +0.00b | 141.00ms | -14.00ms 🚀 |
| dist/workerHasher.1e706e77.js | 11.75kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/EmojiTypeAheadComponent.1990d771.js | 8.67kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/component.77c0f682.js | 6.05kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/card.383bd9bd.js | 5.61kb | +0.00b | 141.00ms | -14.00ms 🚀 |
| dist/media-viewer.59c17dad.js | 3.92kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/EmojiPickerComponent.627aca6b.js | 3.58kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/index.ac562c80.css | 3.46kb | +0.00b | 139.00ms | -11.00ms 🚀 |
| dist/png-chunks-extract.a844fdb6.js | 3.41kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/dropzone.3c67f0a6.js | 3.23kb | +0.00b | 140.00ms | -14.00ms 🚀 |
| dist/Modal.0304b9ed.js | 3.08kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/clipboard.da112b6e.js | 2.75kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/16.290b4a43.js | 2.38kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/EmojiUploadComponent.d54a1fc6.js | 2.18kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/card.eea6ad8c.js | 2.03kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/ResourcedEmojiComponent.abe8596c.js | 2.00kb | +0.00b | 146.00ms | -20.00ms 🚀 |
| dist/date.231500d2.js | 1.73kb | +0.00b | 143.00ms | -14.00ms 🚀 |
| dist/16.d16f14ed.js | 1.73kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/images.80c92f5e.js | 1.67kb | +0.00b | 144.00ms | -13.00ms 🚀 |
| dist/16.1b6ca79f.js | 1.66kb | +0.00b | 145.00ms | -20.00ms 🚀 |
| dist/feedback.d5590b34.js | 1.64kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/browser.75813db0.js | 1.61kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/workerHasher.6d5b22d7.js | 1.54kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/workerHasher.fa198233.js | 1.54kb | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/list-number.c35f9589.js | 1.46kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/status.8ab4226b.js | 1.46kb | +0.00b | 142.00ms | -14.00ms 🚀 |
| dist/16.9ccb5b28.js | 1.38kb | +0.00b | 145.00ms | -12.00ms 🚀 |
| dist/code.3dd71280.js | 1.38kb | +0.00b | 143.00ms | -14.00ms 🚀 |
| dist/16.19c93285.js | 1.33kb | +0.00b | 145.00ms | -20.00ms 🚀 |
| dist/16.c7751e34.js | 1.33kb | +0.00b | 145.00ms | -20.00ms 🚀 |
| dist/16.51049e2c.js | 1.32kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/16.18344f47.js | 1.31kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/link.2f4427e6.js | 1.30kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/heading6.fdbbdf4d.js | 1.30kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/heading3.2113b7a1.js | 1.29kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/16.2c4c1846.js | 1.27kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/emoji.93ae61d3.js | 1.23kb | +0.00b | 144.00ms | -13.00ms 🚀 |
| dist/16.d17ab186.js | 1.23kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/16.436f097f.js | 1.21kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/16.b4da0acb.js | 1.19kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/16.02c68c4f.js | 1.18kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/heading5.e195aab1.js | 1.17kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/expand.450f3aea.js | 1.16kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/media-card-analytics-error-boundary.2323af57.js | 1.15kb | +0.00b | 140.00ms | -14.00ms 🚀 |
| dist/16.b46ef690.js | 1.14kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/16.4e3b0235.js | 1.14kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/16.9ba230ea.js | 1.13kb | +0.00b | 145.00ms | -20.00ms 🚀 |
| dist/16.a37df53b.js | 1.13kb | +0.00b | 145.00ms | -20.00ms 🚀 |
| dist/16.9adb4ec7.js | 1.13kb | +0.00b | 145.00ms | -12.00ms 🚀 |
| dist/16.eafe1a11.js | 1.13kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/heading2.a2a1be3f.js | 1.11kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/16.3920766f.js | 1.10kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/component.6600f83e.js | 1.07kb | +0.00b | 146.00ms | -19.00ms 🚀 |
| dist/mention.4d64e059.js | 1.07kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/heading4.ec980e93.js | 1.06kb | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/16.afcb09b6.js | 1.05kb | +0.00b | 145.00ms | -11.00ms 🚀 |
| dist/layout.52ba15f7.js | 1.05kb | +0.00b | 143.00ms | -14.00ms 🚀 |
| dist/divider.16defde6.js | 1.03kb | +0.00b | 144.00ms | -13.00ms 🚀 |
| dist/quote.aa25b83c.js | 1.02kb | +0.00b | 142.00ms | -14.00ms 🚀 |
| dist/action.9aa72b39.js | 1.00kb | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/decision.20244046.js | 1010.00b | +0.00b | 143.00ms | -14.00ms 🚀 |
| dist/panel-warning.97a0e3b3.js | 1008.00b | +0.00b | 142.00ms | -14.00ms 🚀 |
| dist/list.c9147a62.js | 978.00b | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/heading1.7d899ba5.js | 976.00b | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/panel-error.8d11e8b7.js | 906.00b | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/panel.a28c56ee.js | 897.00b | +0.00b | 142.00ms | -14.00ms 🚀 |
| dist/table.5a4874ee.js | 889.00b | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/panel-success.9b923e58.js | 845.00b | +0.00b | 142.00ms | -14.00ms 🚀 |
| dist/panel-note.97a4d541.js | 841.00b | +0.00b | 143.00ms | -13.00ms 🚀 |
| dist/media-picker-analytics-error-boundary.f8ddfee3.js | 839.00b | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/media-card-analytics-error-boundary.663369b9.js | 835.00b | +0.00b | 141.00ms | -14.00ms 🚀 |
| dist/media-card-analytics-error-boundary.73ba4074.js | 835.00b | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/media-viewer-analytics-error-boundary.0afa7031.js | 755.00b | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/EmojiPickerComponent.2e54b9c6.js | 535.00b | +0.00b | 142.00ms | -13.00ms 🚀 |
| dist/simpleHasher.16c8bf41.js | 528.00b | +0.00b | 141.00ms | -14.00ms 🚀 |
| dist/simpleHasher.a29b2a8e.js | 528.00b | +0.00b | 141.00ms | -13.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 155.00ms | -36.00ms 🚀 |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/276075947201921548)**

mischnic commented on packages/core/core/src/UncommittedAsset.js:
```diff
@@ -124,6 +131,37 @@ export default class UncommittedAsset {
+      [...this.value.includedFiles.values()].map(async file => {
+        if (file.hash == null) {
+          // $FlowFixMe
+          file.hash = await md5FromFilePath(
```
Mutating the object 😬

padmaia commented:
> An alternative would be to somehow send the environment variables and included files from the request graph to the transformer when available, and have it use these values as part of the cache key. But unfortunately, this would invalidate more than needed potentially, because intermediary assets that are created during transformation are not stored in the graph, and are only available in the cache itself.

I think we could make this work. Any env variables used in any of the pipelines would need to invalidate the entire request, but if we attach the env variables only to the assets that use them and use that for the pipeline cache key then we would still be able to use cached intermediate assets. We would just need a way for child assets to not necessarily inherit attached env variables from the parent asset. I think getting rid of the idea of child assets and moving to inline dependencies like we've talked about would make this easier.

devongovett commented:
> I think getting rid of the idea of child assets and moving to inline dependencies like we've talked about would make this easier.

mmm yeah possibly. Though this also happens even when the asset doesn't create more than one child, e.g. TS -> JS. The intermediary JS asset produced by the TS pipeline is cached too...

devongovett commented on packages/core/core/src/AssetGraphBuilder.js:
```diff
@@ -293,6 +294,21 @@ export default class AssetGraphBuilder extends EventEmitter {
+    // Add invalidations to the request if a node already exists in the graph.
+    // These are used to compute the cache key for assets during transformation.
+    if (this.requestGraph.hasNode(request.id)) {
+      request.input.invalidations = this.requestGraph
```
Unsure about mutating the request input here. I imagine this would also get stored in the graph as part of the request. @padmaia do you think we need a way to send arguments to the request function without storing them?

padmaia replied:
Yeah, this feels a bit wrong. And yes, it would end up getting stored in the graph, though we are storing inputs in the graph without much benefit right now. I added that functionality when I was thinking that we might want to rerun all invalidated in requests in parallel, but what I've found is that it's much simpler and often more performant to just start at the top of the request graph again and skip requests that are still valid because requests further up the graph will end up deleting requests further down. So maybe we should think about no longer storing inputs, but that doesn't really solve the problem of needing to pass the invalidations in to the Transformation.

One idea I've had is that we could pass previous results into requests, which would give us the option of iterating on the previous result instead of completely rebuilding it. Was thinking this type of behavior would be useful for an `AssetGraphRequest` once we get there. I know we are trying to keep the graph thin, but since results are being cached we could maybe just store the cache key(s). Might take some refactoring in Transformation, but I think this type of approach would make the code less confusing.

devongovett replied:
The easiest option would be to pass the request graph into request handlers. Then the asset request itself could query the graph for invalidations.

devongovett replied:
Updated to do that: ab81fe8.

padmaia replied:
Requests would be able to do some unsafe things with full access to the graph. What if you added the `getInvalidations` function to the API to passed into requests?

devongovett replied:
Great idea! Done. 😄

devongovett commented on packages/core/core/src/assetUtils.js:
```diff
@@ -194,3 +203,41 @@ export async function getConfig(
+    switch (invalidation.type) {
+      case 'file':
+        hash.update(
+          await md5FromFilePath(options.inputFS, invalidation.filePath),
```
We could probably cache these hashes somewhere at least during a transformation run so we aren't recomputing them all the time if it becomes a perf problem...

devongovett commented on packages/core/integration-tests/test/postcss.js:
```diff
@@ -291,9 +291,6 @@ describe('postcss', () => {
       {
         name: 'style.css',
         assets: ['style.css'],
-        includedFiles: {
```
These aren't attached to asset objects anymore, so there isn't a good way to test this without access to the request graph. These tests weren't really useful anyway since they didn't assert that the cache was invalidated properly, just that the file was listed.

mischnic replied:
I found them useful as a kind of unit test for transformers, as testing this list would be enough under the assumption that core handled it correctly (so that we don't need slow/complicated cache tests for every plugin).

devongovett replied:
Yeah. I think we'll find a way to bring them back. We need a way for the tests to access the internal bundle graph I guess...

padmaia replied:
This kind of felt like testing implementation details anyway. I've been thinking it would be nice to expose the RequestGraph and requests in test utils somehow so transformer plugins could have tests that only run an asset request for instance.

devongovett commented on packages/core/types/index.js:
```diff
@@ -505,8 +504,9 @@ export interface MutableAsset extends BaseAsset {
 
   addDependency(dep: DependencyOptions): string;
-  addIncludedFile(file: File): void;
+  addIncludedFile(filePath: FilePath): void;
```
Debating renaming this to `invalidateOnFileChange` or something. I guess it's different from `invalidateOnFileUpdate` because it also invalidates on delete, and I don't think we need to make plugin authors specify both manually. Thoughts?

padmaia replied:
Yeah I could see a function that sets up both invalidations being useful. Not sure if "change" totally communicates that, I could see people getting "update" and "change" mixed up, but I can't really think of a better word, so I think I'm okay with it :)

devongovett replied:
Ok, let's add this to our API audit separately from this PR

devongovett commented:
Updated as discussed to update the cache key based on the collected invalidations. Also changed the AssetValue type to not hold these as they don't need to be stored in the graph per asset, but rather per request. There is now a single list of "invalidations" rather than an included files list and a env list. This type is also used to send invalidations from the graph to the transformation as part of the request. There's a function to compute a hash based on a list of invalidations for use in the cache key.

padmaia reviewed (approved):
Looks good!
